component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,272 @@
+        if (StringUtils.isEmpty(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+        }
+        if (TokenExchangeConstants.JWT_TOKEN_TYPE.equals(requested_token_type)) {
+            signedJWT = TokenExchangeUtils.getSignedJWT(requestParams.get(TokenExchangeConstants.SUBJECT_TOKEN)[0]);
+            if (signedJWT == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "No Valid subject token was found for "
+                        + TokenExchangeConstants.TOKEN_EXCHANGE_GRANT_TYPE);
+            } else {
+                claimsSet = TokenExchangeUtils.getClaimSet(signedJWT);
+            }
+
+            if (claimsSet == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Claim values are empty in the given JSON Web Token");
+            }
+
+            String jwtIssuer = claimsSet.getIssuer();
+            String subject = resolveSubject(claimsSet);
+            List<String> audiences = claimsSet.getAudience();
+            Date expirationTime = claimsSet.getExpirationTime();
+            Date notBeforeTime = claimsSet.getNotBeforeTime();
+            Date issuedAtTime = claimsSet.getIssueTime();
+            Map<String, Object> customClaims = new HashMap<>(claimsSet.getClaims());
+
+            tokReqMsgCtx.addProperty(TokenExchangeConstants.EXPIRY_TIME, expirationTime);
+            long currentTimeInMillis = System.currentTimeMillis();
+            long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
+
+            if (StringUtils.isEmpty(jwtIssuer) || StringUtils.isEmpty(subject) || expirationTime == null ||
+                    audiences == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Mandatory fields(Issuer, Subject, Expiration time "
+                        + "or Audience) are empty in the given JWT");
+            }
+            identityProvider = TokenExchangeUtils.getIdPByIssuer(jwtIssuer, tenantDomain);
+            tokenEndPointAlias = TokenExchangeUtils.getTokenEndpointAlias(identityProvider, tenantDomain);
+            try {
+                if (signedJWT != null) {
+                    if (TokenExchangeUtils.validateSignature(signedJWT, identityProvider, tenantDomain)) {
+                        log.debug("Signature/MAC validated successfully.");
+                    } else {
+                        handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Signature or Message Authentication " +
+                                "invalid");
+                    }
+                }
+            } catch (JOSEException e) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Error when verifying signature");
+            }
+            TokenExchangeUtils.setAuthorizedUser(tokReqMsgCtx, identityProvider, subject);
+
+            log.debug("Subject(sub) found in JWT: " + subject);
+            log.debug(subject + " set as the Authorized User.");
```

Comment:
shall we check is log debug enabled here as we're doing string concatination

Reply:
Can we have these both messages in a same log and check for debug enabled

Reply:
Fixed in 8693d5454cb05015e9d9614c56a1d09bdf8a8997